### PR TITLE
Fix mis-optimization in predicate push down

### PR DESCRIPTION
**Pull request by rongrong**
This will fixe #13009

It does come with potential performance regression though.

### CODE REVIEW

**Comment by rongrong** [comment]
If you take a look at the test failures, you can get an idea of what kind of regressions we'd have. We can also be more sophisticated in deciding what's ok vs not ok to do in `EqualityInference` though I don't really want to get into the business of enumerating all possible expressions. Another way is to just leave this bug as a feature (not fixing it). The failed query can be fixed by adding a `TRY` anyways. And it does allow the future subfield pushdown to take effect.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java [hunk 1]
```diff
@@ -120,6 +120,16 @@ private Expression rewriteExpression(Expression expression, Predicate<VariableRe
             subExpressions = filter(subExpressions, not(equalTo(expression)));
         }
 
+        if (!variableToExpressionPredicate(variableScope, types).apply(expression)) {
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java]: Do we wanna have a session property to guard this?

Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java]: We can. Putting this PR up to start conversation on how exactly we want to go forward with it. If we want to go with this strict version we definitely should add a session property.

**Comment by rongrong** [comment]
> I have a couple concerns:
> 
> 1. Those performance regressions can be very significant.  For example I've seen queries that only specify the partitioning key for one side of the join, and then add a join condition on that column.  Those predicates would no longer be pushed down after this change and we'd scan the whole table.
That's my motivation of leaving it as is. The potential regression can be huge.
> 2. I'm not 100% convinced that this solves all similar such issues with predicate push down.  Is it ever possible that this would happen without a join, where some predicate that's supposed to execute first gets executed second (e.g. relying on some feature of a non-deterministic function)
> 
The fix is in `EqualityInference` not specific to join, so if similar situations in other nodes it should fix those. Determinism is a different issue though. I think `PredicatePushdown` filters out non-deterministic expressions before considering pushdown (at least in the join code path, I didn't check implementation of all nodes).
> My understanding is that it's a known issue that predicate pushdown isn't careful about dependencies between predicates and we intentionally opted to get the performance improvements at the expense of these occasional errors. I agree that we should fix it, but given that it's a longstanding issue I would prefer to take the time to find a proper solution that will still allow predicates to be pushed down when it's safe (which is most of the time).

There is a safer but not very "good looking" way of fixing this, which is to only optimize for "simple expression".

**Comment by rschlussel** [comment]
To clarify what i meant by relying on features of non-deterministic functions:
Let's say there's a function f(x) that fails on negative input.
If you had a filter `where x > rand() and f(x) = 5`, would we push down f(x)=5 because it's deterministic, but not x>rand() because it's non-deterministic? In that case we could also see failures because we were relying on x > rand() to run first and filter out all the negative numbers.

Anyway, my main concern is about performance.  Since it's a long-standing issue I think it's worth leaving it open and taking the time to do a more complex fix with less substantial performance implications.  https://github.com/prestodb/presto/issues/13009#issuecomment-508204236.

**Pull request closed without merging**